File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dict = {}
        res = []
        for num in nums2:
            if num in dict:
                dict[num] += 1
            else:
                dict[num] = 1
                
        for num in nums1:
            if num in dict and dict[num] != 0 :
                res.append(num)
                dict[num] -= 1
        
        return res
                
            
            

File: 0350-intersection-of-two-arrays-ii/0350-intersection-of-two-arrays-ii.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        
        HashMap<Integer, Integer> map = new HashMap<>();
        ArrayList<Integer> res = new ArrayList<>();
        for(int num:nums1){
            map.put(num, map.getOrDefault(num,0)+1);
        }

        for(int num:nums2){
            if (map.containsKey(num) && map.get(num)>0){
                res.add(num);
                map.put(num, map.get(num)-1);
            }
        }
        int[] output = new int[res.size()];
        for (int i = 0; i< res.size(); i++){
            output[i] = res.get(i);
        }
        return output;
    }
}